# 1
Import the tools.

In [4]:
import pandas as pd
import numpy as np
import copy

# 2
Read the data.

In [5]:
flights = pd.read_csv('C:\\Users\\utente\\Documents\\DSSC_IIyear\\Voli\\Dati\\flights.csv')
sectorhours = pd.read_csv('C:\\Users\\utente\\Documents\\DSSC_IIyear\\Voli\\Dati\\sectorhours.csv')
trajectories = pd.read_csv('C:\\Users\\utente\\Documents\\DSSC_IIyear\\Voli\\Dati\\trajectories.csv')

# 3
Print the data.

In [6]:
print(flights.columns)
print (sectorhours.columns)
print (trajectories.columns)
flights
sectorhours
trajectories

Index(['uuid', 'O_D_A', 'dep_time', 'original_trajectory', 'dep_time_minutes'], dtype='object')
Index(['sector_hours', 'icao', 'role', 'minute_start', 'minute_end',
       'capacity'],
      dtype='object')
Index(['centroid', 'eobt', 'icao', 'time_entry', 'time_to_sector_from_origin'], dtype='object')


centroid                 eobt      icao           time_entry  \
0        UKLL_UKDD_AT72_2  2017-09-05 15:41:00  UKLVJEWH  2017-09-05 16:01:32   
1        UKLL_UKDD_AT72_2  2017-09-05 15:41:00  UKBVO355  2017-09-05 16:34:35   
2        UKLL_UKDD_AT72_2  2017-09-05 15:41:00    UKBVSO  2017-09-05 16:34:35   
3        UKLL_UKDD_AT72_2  2017-09-05 15:41:00   UKBVTC1  2017-09-05 16:54:24   
4        UKLL_UKDD_AT72_2  2017-09-05 15:41:00  UKBVTC12  2017-09-05 16:54:24   
...                   ...                  ...       ...                  ...   
2466872  LWSK_EDDF_AT43_1  2017-09-01 13:00:00   EDGGPSA  2017-09-01 14:52:59   
2466873  LWSK_EDDF_AT43_1  2017-09-01 13:00:00   EDGGADF  2017-09-01 14:53:54   
2466874  LMML_HLLB_OTHR_1  2017-02-16 21:45:00   LMMMWST  2017-02-16 22:11:47   
2466875  LMML_HLLB_OTHR_1  2017-02-16 21:45:00   LMMMALL  2017-02-16 22:11:47   
2466876  LMML_HLLB_OTHR_1  2017-02-16 21:45:00   HLLLALL  2017-02-16 22:39:10   

         time_to_sector_from_origin  
0                                20  
1                                53  
2                                53  
3                                73  
4                                73  
...                             ...  
2466872                         112  
2466873                         113  
2466874                          26  
2466875                          26  
2466876                          54  

[2466877 rows x 5 columns]

# 4
Drop unnecessary columns in "flights".

In [7]:
flights = flights.drop(columns=['original_trajectory', 'dep_time'])
flights

uuid           O_D_A  dep_time_minutes
0      GRL780_BGSF_EKCH_20170901134000  BGSF_EKCH_A332               830
1      N174JF_BGSF_LOWI_20170901103500  BGSF_LOWI_AT43               645
2      WOW462_BIKF_EBBR_20170901065500  BIKF_EBBR_A321               385
3      ICE554_BIKF_EBBR_20170901074000  BIKF_EBBR_B752               470
4      WOW720_BIKF_EDDB_20170901054500  BIKF_EDDB_A321               355
...                                ...             ...               ...
29912   HVN31_VVTS_EDDF_20170901155500  VVTS_EDDF_A332               965
29913   HVN51_VVTS_EGLL_20170901173500  VVTS_EGLL_A332              1065
29914  AFR253_VVTS_LFPG_20170901013000  VVTS_LFPG_A332               100
29915  THY166_VVTS_LTBA_20170901144000  VVTS_LTBA_A332               890
29916  BWJ998_ZYHB_EGGW_20170901110000  ZYHB_EGGW_E190               670

[29917 rows x 3 columns]

# 5
Consider only the sectorhours with role 'E'. <br>Reset the index.

In [8]:
sectorhours = sectorhours[sectorhours['role']=='E']
sectorhours.reset_index(drop=True, inplace=True)
sectorhours

sector_hours      icao role  minute_start  minute_end  capacity
0       EGTTOCK_0000_0100_E   EGTTOCK    E             0          59        49
1       EGTTOCK_0100_0200_E   EGTTOCK    E            60         119        49
2       EGTTOCK_0200_0300_E   EGTTOCK    E           120         179        49
3       EGTTOCK_0300_0400_E   EGTTOCK    E           180         239        49
4       EGTTOCK_0400_0500_E   EGTTOCK    E           240         299        49
...                     ...       ...  ...           ...         ...       ...
13485   LSAGL67_2010_2040_E   LSAGL67    E          1210        1239        23
13486     LYBAL_1659_1800_E     LYBAL    E          1019        1079        41
13487    LYBATN_1600_1700_E    LYBATN    E           960        1019        43
13488    LYBAUN_1600_1700_E    LYBAUN    E           960        1019        43
13489  LYBAUWES_1659_1800_E  LYBAUWES    E          1019        1079        54

[13490 rows x 6 columns]

# 6
Create a new column 'O_D_A' in trajectories, obtained from the first 14 characters of 'centroid'.
<br> Drop the unnecessary columns.
<br> Rename the columnns.
<br> Check that the entries of 'O_D_A' are all of lenght 14. 

In [9]:
trajectories['O_D_A'] = trajectories['centroid'].str[0:14]
trajectories = trajectories.drop(columns=['eobt', 'time_entry'])
trajectories = trajectories[['O_D_A', 'centroid', 'icao', 'time_to_sector_from_origin']]

In [10]:
print(len(trajectories))
print(len(trajectories[(trajectories['O_D_A']).str.len()==14]))
print(len(trajectories[(trajectories['O_D_A']).str.len()!=14]))
trajectories

2466877
2466877
0


O_D_A          centroid      icao  \
0        UKLL_UKDD_AT72  UKLL_UKDD_AT72_2  UKLVJEWH   
1        UKLL_UKDD_AT72  UKLL_UKDD_AT72_2  UKBVO355   
2        UKLL_UKDD_AT72  UKLL_UKDD_AT72_2    UKBVSO   
3        UKLL_UKDD_AT72  UKLL_UKDD_AT72_2   UKBVTC1   
4        UKLL_UKDD_AT72  UKLL_UKDD_AT72_2  UKBVTC12   
...                 ...               ...       ...   
2466872  LWSK_EDDF_AT43  LWSK_EDDF_AT43_1   EDGGPSA   
2466873  LWSK_EDDF_AT43  LWSK_EDDF_AT43_1   EDGGADF   
2466874  LMML_HLLB_OTHR  LMML_HLLB_OTHR_1   LMMMWST   
2466875  LMML_HLLB_OTHR  LMML_HLLB_OTHR_1   LMMMALL   
2466876  LMML_HLLB_OTHR  LMML_HLLB_OTHR_1   HLLLALL   

         time_to_sector_from_origin  
0                                20  
1                                53  
2                                53  
3                                73  
4                                73  
...                             ...  
2466872                         112  
2466873                         113  
2466874                          26  
2466875                          26  
2466876                          54  

[2466877 rows x 4 columns]

In [19]:
len(trajectories['O_D_A'].unique())

18691

In [17]:
traj = trajectories.groupby('O_D_A')
traj = traj['centroid'].unique()
traj = pd.DataFrame({'O_D_A':traj.index, 'trajectories':traj.values})
traj

for i in range(0,len(traj)):
    traj['trajectories'].loc[i] = traj['trajectories'].loc[i].tolist()

len(traj['trajectories'].loc[2])

traj['len'] = (traj['trajectories']).apply(len)

In [20]:
len(traj['trajectories'].loc[2])
traj

O_D_A                                       trajectories  len
0      BGSF_EKCH_A332  [BGSF_EKCH_A332_5, BGSF_EKCH_A332_7, BGSF_EKCH...   10
1      BGSF_LOWI_AT43                                 [BGSF_LOWI_AT43_1]    1
2      BIKF_EBBR_A321               [BIKF_EBBR_A321_1, BIKF_EBBR_A321_2]    2
3      BIKF_EBBR_B752  [BIKF_EBBR_B752_2, BIKF_EBBR_B752_4, BIKF_EBBR...    5
4      BIKF_EDDB_A321  [BIKF_EDDB_A321_6, BIKF_EDDB_A321_1, BIKF_EDDB...    6
...               ...                                                ...  ...
18686  VVTS_EDDF_A332  [VVTS_EDDF_A332_3, VVTS_EDDF_A332_5, VVTS_EDDF...    5
18687  VVTS_EGLL_A332  [VVTS_EGLL_A332_1, VVTS_EGLL_A332_3, VVTS_EGLL...    3
18688  VVTS_LFPG_A332  [VVTS_LFPG_A332_6, VVTS_LFPG_A332_3, VVTS_LFPG...    6
18689  VVTS_LTBA_A332               [VVTS_LTBA_A332_2, VVTS_LTBA_A332_1]    2
18690  ZYHB_EGGW_E190                                 [ZYHB_EGGW_E190_1]    1

[18691 rows x 3 columns]

In [21]:
len_dict = dict(zip(traj.O_D_A, traj.len))
len_dict

{'BGSF_EKCH_A332': 10,
 'BGSF_LOWI_AT43': 1,
 'BIKF_EBBR_A321': 2,
 'BIKF_EBBR_B752': 5,
 'BIKF_EDDB_A321': 6,
 'BIKF_EDDF_A321': 13,
 'BIKF_EDDF_B752': 15,
 'BIKF_EDDH_A319': 1,
 'BIKF_EDDH_B752': 5,
 'BIKF_EDDK_A319': 1,
 'BIKF_EDDL_A320': 13,
 'BIKF_EDDM_A320': 3,
 'BIKF_EDDM_B752': 8,
 'BIKF_EDDS_A319': 1,
 'BIKF_EDDT_A319': 2,
 'BIKF_EDDT_A320': 7,
 'BIKF_EFHK_A321': 7,
 'BIKF_EFHK_B752': 4,
 'BIKF_EFHK_B763': 2,
 'BIKF_EGCC_B752': 5,
 'BIKF_EGGD_A320': 4,
 'BIKF_EGGW_A320': 4,
 'BIKF_EGKK_A320': 7,
 'BIKF_EGKK_A321': 8,
 'BIKF_EGKK_B752': 8,
 'BIKF_EGLL_A320': 1,
 'BIKF_EGLL_B752': 12,
 'BIKF_EGLL_B763': 8,
 'BIKF_EGNX_B752': 3,
 'BIKF_EGPD_DH8D': 2,
 'BIKF_EGPF_B752': 3,
 'BIKF_EGPH_A321': 3,
 'BIKF_EHAM_A321': 9,
 'BIKF_EHAM_B752': 7,
 'BIKF_EHAM_B763': 3,
 'BIKF_EICK_A320': 1,
 'BIKF_EIDW_A321': 11,
 'BIKF_EIDW_B734': 8,
 'BIKF_EIME_OTHR': 1,
 'BIKF_EKBI_B752': 4,
 'BIKF_EKCH_A320': 10,
 'BIKF_EKCH_A321': 4,
 'BIKF_EKCH_B752': 2,
 'BIKF_ENBR_B738': 1,
 'BIKF_ENBR_B752': 3,
 'B

In [22]:
len_dict['BIKF_EFHK_A321']

7

# 7
Merge "trajectories" with "flights".

E' indifferente applicare merge con modalità 'inner' o 'right': tutte le traiettorie in flights hanno un corrispondente in trajectories (il viceversa non è vero, come si potrebbe verificare applicando la modalità 'left': ci sono traiettorie in trajectories che non sono state volate il 1/9/2017).

In [23]:
merged = trajectories.merge(flights, how='right', on = 'O_D_A')
merged

O_D_A          centroid      icao  \
0        UKLL_UKDD_AT72  UKLL_UKDD_AT72_2  UKLVJEWH   
1        UKLL_UKDD_AT72  UKLL_UKDD_AT72_2  UKBVO355   
2        UKLL_UKDD_AT72  UKLL_UKDD_AT72_2    UKBVSO   
3        UKLL_UKDD_AT72  UKLL_UKDD_AT72_2   UKBVTC1   
4        UKLL_UKDD_AT72  UKLL_UKDD_AT72_2  UKBVTC12   
...                 ...               ...       ...   
4135268  LWSK_EDDF_AT43  LWSK_EDDF_AT43_1   EDGGKTG   
4135269  LWSK_EDDF_AT43  LWSK_EDDF_AT43_1     EDGG4   
4135270  LWSK_EDDF_AT43  LWSK_EDDF_AT43_1  EDGGKIHA   
4135271  LWSK_EDDF_AT43  LWSK_EDDF_AT43_1   EDGGPSA   
4135272  LWSK_EDDF_AT43  LWSK_EDDF_AT43_1   EDGGADF   

         time_to_sector_from_origin                             uuid  \
0                                20  WRC124_UKLL_UKDD_20170901180000   
1                                53  WRC124_UKLL_UKDD_20170901180000   
2                                53  WRC124_UKLL_UKDD_20170901180000   
3                                73  WRC124_UKLL_UKDD_20170901180000   
4                                73  WRC124_UKLL_UKDD_20170901180000   
...                             ...                              ...   
4135268                         104   Z3MKD_LWSK_EDDF_20170901130000   
4135269                         104   Z3MKD_LWSK_EDDF_20170901130000   
4135270                         104   Z3MKD_LWSK_EDDF_20170901130000   
4135271                         112   Z3MKD_LWSK_EDDF_20170901130000   
4135272                         113   Z3MKD_LWSK_EDDF_20170901130000   

         dep_time_minutes  
0                    1090  
1                    1090  
2                    1090  
3                    1090  
4                    1090  
...                   ...  
4135268               785  
4135269               785  
4135270               785  
4135271               785  
4135272               785  

[4135273 rows x 6 columns]

# 8
Drop unnecessary columns.
<br>Reorder the columns.
<br>Rename the columns.

In [24]:
merged = merged[['uuid', 'O_D_A', 'centroid', 'icao', 'dep_time_minutes', 'time_to_sector_from_origin']]
merged.columns = ['flight', 'O_D_A', 'trajectory', 'icao', 'dep_time', 'time_to_sector']
merged

flight           O_D_A        trajectory  \
0        WRC124_UKLL_UKDD_20170901180000  UKLL_UKDD_AT72  UKLL_UKDD_AT72_2   
1        WRC124_UKLL_UKDD_20170901180000  UKLL_UKDD_AT72  UKLL_UKDD_AT72_2   
2        WRC124_UKLL_UKDD_20170901180000  UKLL_UKDD_AT72  UKLL_UKDD_AT72_2   
3        WRC124_UKLL_UKDD_20170901180000  UKLL_UKDD_AT72  UKLL_UKDD_AT72_2   
4        WRC124_UKLL_UKDD_20170901180000  UKLL_UKDD_AT72  UKLL_UKDD_AT72_2   
...                                  ...             ...               ...   
4135268   Z3MKD_LWSK_EDDF_20170901130000  LWSK_EDDF_AT43  LWSK_EDDF_AT43_1   
4135269   Z3MKD_LWSK_EDDF_20170901130000  LWSK_EDDF_AT43  LWSK_EDDF_AT43_1   
4135270   Z3MKD_LWSK_EDDF_20170901130000  LWSK_EDDF_AT43  LWSK_EDDF_AT43_1   
4135271   Z3MKD_LWSK_EDDF_20170901130000  LWSK_EDDF_AT43  LWSK_EDDF_AT43_1   
4135272   Z3MKD_LWSK_EDDF_20170901130000  LWSK_EDDF_AT43  LWSK_EDDF_AT43_1   

             icao  dep_time  time_to_sector  
0        UKLVJEWH      1090              20  
1        UKBVO355      1090              53  
2          UKBVSO      1090              53  
3         UKBVTC1      1090              73  
4        UKBVTC12      1090              73  
...           ...       ...             ...  
4135268   EDGGKTG       785             104  
4135269     EDGG4       785             104  
4135270  EDGGKIHA       785             104  
4135271   EDGGPSA       785             112  
4135272   EDGGADF       785             113  

[4135273 rows x 6 columns]

# 9
Create a new column 'entry', describing the time a flight enters a sector-hour.
<br>The modulo operation is needed since we might exceed midnight (60x24=1440).

In [25]:
merged['entry'] = (merged['dep_time'] + merged['time_to_sector'])%1440
merged

C:\Users\utente\AppData\Local\Programs\Python\Python37\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


flight           O_D_A        trajectory  \
0        WRC124_UKLL_UKDD_20170901180000  UKLL_UKDD_AT72  UKLL_UKDD_AT72_2   
1        WRC124_UKLL_UKDD_20170901180000  UKLL_UKDD_AT72  UKLL_UKDD_AT72_2   
2        WRC124_UKLL_UKDD_20170901180000  UKLL_UKDD_AT72  UKLL_UKDD_AT72_2   
3        WRC124_UKLL_UKDD_20170901180000  UKLL_UKDD_AT72  UKLL_UKDD_AT72_2   
4        WRC124_UKLL_UKDD_20170901180000  UKLL_UKDD_AT72  UKLL_UKDD_AT72_2   
...                                  ...             ...               ...   
4135268   Z3MKD_LWSK_EDDF_20170901130000  LWSK_EDDF_AT43  LWSK_EDDF_AT43_1   
4135269   Z3MKD_LWSK_EDDF_20170901130000  LWSK_EDDF_AT43  LWSK_EDDF_AT43_1   
4135270   Z3MKD_LWSK_EDDF_20170901130000  LWSK_EDDF_AT43  LWSK_EDDF_AT43_1   
4135271   Z3MKD_LWSK_EDDF_20170901130000  LWSK_EDDF_AT43  LWSK_EDDF_AT43_1   
4135272   Z3MKD_LWSK_EDDF_20170901130000  LWSK_EDDF_AT43  LWSK_EDDF_AT43_1   

             icao  dep_time  time_to_sector  entry  
0        UKLVJEWH      1090              20   1110  
1        UKBVO355      1090              53   1143  
2          UKBVSO      1090              53   1143  
3         UKBVTC1      1090              73   1163  
4        UKBVTC12      1090              73   1163  
...           ...       ...             ...    ...  
4135268   EDGGKTG       785             104    889  
4135269     EDGG4       785             104    889  
4135270  EDGGKIHA       785             104    889  
4135271   EDGGPSA       785             112    897  
4135272   EDGGADF       785             113    898  

[4135273 rows x 7 columns]

# 10
Check that the maximum value of the column 'entry' doesn't exceed 1440.

In [26]:
print ("Is the maximum value of the column 'entry' smaller than 1440?")
print (max(merged['entry'])<1440)
print("max(merged['entry']) =", max(merged['entry']))

Is the maximum value of the column 'entry' smaller than 1440?
True
max(merged['entry']) = 1439


# 11
Merge "merged" with "sectorhours", based on 'icao'.

In [27]:
merged_2 = merged.merge(sectorhours, how='left', left_on='icao', right_on='icao')
merged_2

flight           O_D_A        trajectory  \
0         WRC124_UKLL_UKDD_20170901180000  UKLL_UKDD_AT72  UKLL_UKDD_AT72_2   
1         WRC124_UKLL_UKDD_20170901180000  UKLL_UKDD_AT72  UKLL_UKDD_AT72_2   
2         WRC124_UKLL_UKDD_20170901180000  UKLL_UKDD_AT72  UKLL_UKDD_AT72_2   
3         WRC124_UKLL_UKDD_20170901180000  UKLL_UKDD_AT72  UKLL_UKDD_AT72_2   
4         WRC124_UKLL_UKDD_20170901180000  UKLL_UKDD_AT72  UKLL_UKDD_AT72_2   
...                                   ...             ...               ...   
35353263   Z3MKD_LWSK_EDDF_20170901130000  LWSK_EDDF_AT43  LWSK_EDDF_AT43_1   
35353264   Z3MKD_LWSK_EDDF_20170901130000  LWSK_EDDF_AT43  LWSK_EDDF_AT43_1   
35353265   Z3MKD_LWSK_EDDF_20170901130000  LWSK_EDDF_AT43  LWSK_EDDF_AT43_1   
35353266   Z3MKD_LWSK_EDDF_20170901130000  LWSK_EDDF_AT43  LWSK_EDDF_AT43_1   
35353267   Z3MKD_LWSK_EDDF_20170901130000  LWSK_EDDF_AT43  LWSK_EDDF_AT43_1   

              icao  dep_time  time_to_sector  entry          sector_hours  \
0         UKLVJEWH      1090              20   1110  UKLVJEWH_0000_0100_E   
1         UKLVJEWH      1090              20   1110  UKLVJEWH_0100_0200_E   
2         UKLVJEWH      1090              20   1110  UKLVJEWH_0200_0300_E   
3         UKLVJEWH      1090              20   1110  UKLVJEWH_0300_0400_E   
4         UKLVJEWH      1090              20   1110  UKLVJEWH_0400_0500_E   
...            ...       ...             ...    ...                   ...   
35353263   EDGGPSA       785             112    897   EDGGPSA_1700_1800_E   
35353264   EDGGPSA       785             112    897   EDGGPSA_1800_1900_E   
35353265   EDGGPSA       785             112    897   EDGGPSA_1900_2000_E   
35353266   EDGGPSA       785             112    897   EDGGPSA_2000_2100_E   
35353267   EDGGADF       785             113    898                   NaN   

         role  minute_start  minute_end  capacity  
0           E           0.0        59.0      39.0  
1           E          60.0       119.0      39.0  
2           E         120.0       179.0      39.0  
3           E         180.0       239.0      39.0  
4           E         240.0       299.0      39.0  
...       ...           ...         ...       ...  
35353263    E        1020.0      1079.0      38.0  
35353264    E        1080.0      1139.0      38.0  
35353265    E        1140.0      1199.0      38.0  
35353266    E        1200.0      1259.0      38.0  
35353267  NaN           NaN         NaN       NaN  

[35353268 rows x 12 columns]

The modality 'left' keeps all the sectorhours present in "merged", even when they don't have any corrispondence in "sectorhours" (the relevant entries are set to 'NaN').
The modality 'inner', instead, would keep only the sectorhours in "merged" having an icao which is also present in "sectorhours", thus leaving out a number of flights.

# 12
Keep only the sectorhours having a 'minute_start' not greater than 'entry' +30 and an 'entry' not smaller than 'minute_end' +30.
<br>Compare the number of flights included in "merged_2" and "clean" respectively.
See #18 for the reasoning.

In [61]:
clean = merged_2[(merged_2['minute_start']<merged_2['entry']+30) & (merged_2['entry']<=merged_2['minute_end']+30)]
#clean_prova = merged_2[(merged_2['minute_start']<=merged_2['entry']+29) & (merged_2['entry']<=merged_2['minute_end']+30)]
#print(len(clean)==len(clean_prova))
clean

flight           O_D_A        trajectory  \
18        WRC124_UKLL_UKDD_20170901180000  UKLL_UKDD_AT72  UKLL_UKDD_AT72_2   
43        WRC124_UKLL_UKDD_20170901180000  UKLL_UKDD_AT72  UKLL_UKDD_AT72_2   
44        WRC124_UKLL_UKDD_20170901180000  UKLL_UKDD_AT72  UKLL_UKDD_AT72_2   
67        WRC124_UKLL_UKDD_20170901180000  UKLL_UKDD_AT72  UKLL_UKDD_AT72_2   
68        WRC124_UKLL_UKDD_20170901180000  UKLL_UKDD_AT72  UKLL_UKDD_AT72_2   
...                                   ...             ...               ...   
35353208   Z3MKD_LWSK_EDDF_20170901130000  LWSK_EDDF_AT43  LWSK_EDDF_AT43_1   
35353227   Z3MKD_LWSK_EDDF_20170901130000  LWSK_EDDF_AT43  LWSK_EDDF_AT43_1   
35353228   Z3MKD_LWSK_EDDF_20170901130000  LWSK_EDDF_AT43  LWSK_EDDF_AT43_1   
35353260   Z3MKD_LWSK_EDDF_20170901130000  LWSK_EDDF_AT43  LWSK_EDDF_AT43_1   
35353261   Z3MKD_LWSK_EDDF_20170901130000  LWSK_EDDF_AT43  LWSK_EDDF_AT43_1   

              icao  dep_time  time_to_sector  entry          sector_hours  \
18        UKLVJEWH      1090              20   1110  UKLVJEWH_1800_1900_E   
43          UKBVSO      1090              53   1143    UKBVSO_1800_1900_E   
44          UKBVSO      1090              53   1143    UKBVSO_1900_2000_E   
67        UKBVTC12      1090              73   1163  UKBVTC12_1800_1900_E   
68        UKBVTC12      1090              73   1163  UKBVTC12_1900_2000_E   
...            ...       ...             ...    ...                   ...   
35353208   EDMMALB       785              97    882   EDMMALB_1500_1600_E   
35353227   EDGGKTG       785             104    889   EDGGKTG_1400_1500_E   
35353228   EDGGKTG       785             104    889   EDGGKTG_1500_1600_E   
35353260   EDGGPSA       785             112    897   EDGGPSA_1400_1500_E   
35353261   EDGGPSA       785             112    897   EDGGPSA_1500_1600_E   

         role  minute_start  minute_end  capacity  
18          E        1080.0      1139.0      39.0  
43          E        1080.0      1139.0      36.0  
44          E        1140.0      1199.0      36.0  
67          E        1080.0      1139.0      34.0  
68          E        1140.0      1199.0      34.0  
...       ...           ...         ...       ...  
35353208    E         900.0       959.0      46.0  
35353227    E         840.0       899.0      41.0  
35353228    E         900.0       959.0      41.0  
35353260    E         840.0       899.0      38.0  
35353261    E         900.0       959.0      38.0  

[2882095 rows x 12 columns]

In [62]:
clean[clean['flight']=='WRC124_UKLL_UKDD_20170901180000']

flight           O_D_A        trajectory  \
18   WRC124_UKLL_UKDD_20170901180000  UKLL_UKDD_AT72  UKLL_UKDD_AT72_2   
43   WRC124_UKLL_UKDD_20170901180000  UKLL_UKDD_AT72  UKLL_UKDD_AT72_2   
44   WRC124_UKLL_UKDD_20170901180000  UKLL_UKDD_AT72  UKLL_UKDD_AT72_2   
67   WRC124_UKLL_UKDD_20170901180000  UKLL_UKDD_AT72  UKLL_UKDD_AT72_2   
68   WRC124_UKLL_UKDD_20170901180000  UKLL_UKDD_AT72  UKLL_UKDD_AT72_2   
105  WRC124_UKLL_UKDD_20170901180000  UKLL_UKDD_AT72  UKLL_UKDD_AT72_2   
106  WRC124_UKLL_UKDD_20170901180000  UKLL_UKDD_AT72  UKLL_UKDD_AT72_2   
129  WRC124_UKLL_UKDD_20170901180000  UKLL_UKDD_AT72  UKLL_UKDD_AT72_2   
130  WRC124_UKLL_UKDD_20170901180000  UKLL_UKDD_AT72  UKLL_UKDD_AT72_2   
153  WRC124_UKLL_UKDD_20170901180000  UKLL_UKDD_AT72  UKLL_UKDD_AT72_2   
154  WRC124_UKLL_UKDD_20170901180000  UKLL_UKDD_AT72  UKLL_UKDD_AT72_2   
175  WRC124_UKLL_UKDD_20170901180000  UKLL_UKDD_AT72  UKLL_UKDD_AT72_1   
176  WRC124_UKLL_UKDD_20170901180000  UKLL_UKDD_AT72  UKLL_UKDD_AT72_1   
201  WRC124_UKLL_UKDD_20170901180000  UKLL_UKDD_AT72  UKLL_UKDD_AT72_1   
202  WRC124_UKLL_UKDD_20170901180000  UKLL_UKDD_AT72  UKLL_UKDD_AT72_1   
225  WRC124_UKLL_UKDD_20170901180000  UKLL_UKDD_AT72  UKLL_UKDD_AT72_1   
226  WRC124_UKLL_UKDD_20170901180000  UKLL_UKDD_AT72  UKLL_UKDD_AT72_1   
250  WRC124_UKLL_UKDD_20170901180000  UKLL_UKDD_AT72  UKLL_UKDD_AT72_1   
251  WRC124_UKLL_UKDD_20170901180000  UKLL_UKDD_AT72  UKLL_UKDD_AT72_1   

          icao  dep_time  time_to_sector  entry           sector_hours role  \
18    UKLVJEWH      1090              20   1110   UKLVJEWH_1800_1900_E    E   
43      UKBVSO      1090              53   1143     UKBVSO_1800_1900_E    E   
44      UKBVSO      1090              53   1143     UKBVSO_1900_2000_E    E   
67    UKBVTC12      1090              73   1163   UKBVTC12_1800_1900_E    E   
68    UKBVTC12      1090              73   1163   UKBVTC12_1900_2000_E    E   
105     UKBVNM      1090              91   1181     UKBVNM_1900_2000_E    E   
106     UKBVNM      1090              91   1181     UKBVNM_2000_2100_E    E   
129  UKDVCE325      1090             112   1202  UKDVCE325_1900_2000_E    E   
130  UKDVCE325      1090             112   1202  UKDVCE325_2000_2100_E    E   
153     UKDDTA      1090             123   1213     UKDDTA_1900_2000_E    E   
154     UKDDTA      1090             123   1213     UKDDTA_2000_2100_E    E   
175   UKLVJEWH      1090              17   1107   UKLVJEWH_1700_1800_E    E   
176   UKLVJEWH      1090              17   1107   UKLVJEWH_1800_1900_E    E   
201     UKBVSO      1090              47   1137     UKBVSO_1800_1900_E    E   
202     UKBVSO      1090              47   1137     UKBVSO_1900_2000_E    E   
225  UKDVCE325      1090              79   1169  UKDVCE325_1800_1900_E    E   
226  UKDVCE325      1090              79   1169  UKDVCE325_1900_2000_E    E   
250     UKDDTA      1090              89   1179     UKDDTA_1900_2000_E    E   
251     UKDDTA      1090              89   1179     UKDDTA_2000_2100_E    E   

     minute_start  minute_end  capacity  
18         1080.0      1139.0      39.0  
43         1080.0      1139.0      36.0  
44         1140.0      1199.0      36.0  
67         1080.0      1139.0      34.0  
68         1140.0      1199.0      34.0  
105        1140.0      1199.0      41.0  
106        1200.0      1259.0      41.0  
129        1140.0      1199.0      30.0  
130        1200.0      1259.0      30.0  
153        1140.0      1199.0      30.0  
154        1200.0      1259.0      30.0  
175        1020.0      1079.0      39.0  
176        1080.0      1139.0      39.0  
201        1080.0      1139.0      36.0  
202        1140.0      1199.0      36.0  
225        1080.0      1139.0      30.0  
226        1140.0      1199.0      30.0  
250        1140.0      1199.0      30.0  
251        1200.0      1259.0      30.0

In [63]:
def f_list(row):
    return len_dict[row['O_D_A']]

In [64]:
clean['number'] = clean.apply(f_list, axis=1)

C:\Users\utente\AppData\Local\Programs\Python\Python37\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [65]:
clean

flight           O_D_A        trajectory  \
18        WRC124_UKLL_UKDD_20170901180000  UKLL_UKDD_AT72  UKLL_UKDD_AT72_2   
43        WRC124_UKLL_UKDD_20170901180000  UKLL_UKDD_AT72  UKLL_UKDD_AT72_2   
44        WRC124_UKLL_UKDD_20170901180000  UKLL_UKDD_AT72  UKLL_UKDD_AT72_2   
67        WRC124_UKLL_UKDD_20170901180000  UKLL_UKDD_AT72  UKLL_UKDD_AT72_2   
68        WRC124_UKLL_UKDD_20170901180000  UKLL_UKDD_AT72  UKLL_UKDD_AT72_2   
...                                   ...             ...               ...   
35353208   Z3MKD_LWSK_EDDF_20170901130000  LWSK_EDDF_AT43  LWSK_EDDF_AT43_1   
35353227   Z3MKD_LWSK_EDDF_20170901130000  LWSK_EDDF_AT43  LWSK_EDDF_AT43_1   
35353228   Z3MKD_LWSK_EDDF_20170901130000  LWSK_EDDF_AT43  LWSK_EDDF_AT43_1   
35353260   Z3MKD_LWSK_EDDF_20170901130000  LWSK_EDDF_AT43  LWSK_EDDF_AT43_1   
35353261   Z3MKD_LWSK_EDDF_20170901130000  LWSK_EDDF_AT43  LWSK_EDDF_AT43_1   

              icao  dep_time  time_to_sector  entry          sector_hours  \
18        UKLVJEWH      1090              20   1110  UKLVJEWH_1800_1900_E   
43          UKBVSO      1090              53   1143    UKBVSO_1800_1900_E   
44          UKBVSO      1090              53   1143    UKBVSO_1900_2000_E   
67        UKBVTC12      1090              73   1163  UKBVTC12_1800_1900_E   
68        UKBVTC12      1090              73   1163  UKBVTC12_1900_2000_E   
...            ...       ...             ...    ...                   ...   
35353208   EDMMALB       785              97    882   EDMMALB_1500_1600_E   
35353227   EDGGKTG       785             104    889   EDGGKTG_1400_1500_E   
35353228   EDGGKTG       785             104    889   EDGGKTG_1500_1600_E   
35353260   EDGGPSA       785             112    897   EDGGPSA_1400_1500_E   
35353261   EDGGPSA       785             112    897   EDGGPSA_1500_1600_E   

         role  minute_start  minute_end  capacity  number  
18          E        1080.0      1139.0      39.0       2  
43          E        1080.0      1139.0      36.0       2  
44          E        1140.0      1199.0      36.0       2  
67          E        1080.0      1139.0      34.0       2  
68          E        1140.0      1199.0      34.0       2  
...       ...           ...         ...       ...     ...  
35353208    E         900.0       959.0      46.0       1  
35353227    E         840.0       899.0      41.0       1  
35353228    E         900.0       959.0      41.0       1  
35353260    E         840.0       899.0      38.0       1  
35353261    E         900.0       959.0      38.0       1  

[2882095 rows x 13 columns]

In [66]:
#max(clean['number'])
#clean[clean['number']>60]
#len_dict['RJTT_EDDF_B744']

In [67]:
clean['bool'] = clean['entry'] + 30 >= clean['minute_end']

C:\Users\utente\AppData\Local\Programs\Python\Python37\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [68]:
def f_compute(row):
    if row['bool'] == True:
        value = row['minute_end'] - (row['entry'] - 30) + 1
    else:
        value = row['entry'] + 30 - row['minute_start']
    return max(0,value)

In [69]:
#def prova(b):
#    return max(0,b)

#prova(5)

In [70]:
clean['overlap'] = clean.apply(f_compute, axis=1)

C:\Users\utente\AppData\Local\Programs\Python\Python37\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [71]:
#max(clean['overlap'])
# clean[clean['overlap']>60]

In [72]:
clean['prob'] = clean['overlap']/60

C:\Users\utente\AppData\Local\Programs\Python\Python37\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [73]:
clean['weighted'] = clean['prob']/clean['number']

C:\Users\utente\AppData\Local\Programs\Python\Python37\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [75]:
clean

flight           O_D_A        trajectory  \
18        WRC124_UKLL_UKDD_20170901180000  UKLL_UKDD_AT72  UKLL_UKDD_AT72_2   
43        WRC124_UKLL_UKDD_20170901180000  UKLL_UKDD_AT72  UKLL_UKDD_AT72_2   
44        WRC124_UKLL_UKDD_20170901180000  UKLL_UKDD_AT72  UKLL_UKDD_AT72_2   
67        WRC124_UKLL_UKDD_20170901180000  UKLL_UKDD_AT72  UKLL_UKDD_AT72_2   
68        WRC124_UKLL_UKDD_20170901180000  UKLL_UKDD_AT72  UKLL_UKDD_AT72_2   
...                                   ...             ...               ...   
35353208   Z3MKD_LWSK_EDDF_20170901130000  LWSK_EDDF_AT43  LWSK_EDDF_AT43_1   
35353227   Z3MKD_LWSK_EDDF_20170901130000  LWSK_EDDF_AT43  LWSK_EDDF_AT43_1   
35353228   Z3MKD_LWSK_EDDF_20170901130000  LWSK_EDDF_AT43  LWSK_EDDF_AT43_1   
35353260   Z3MKD_LWSK_EDDF_20170901130000  LWSK_EDDF_AT43  LWSK_EDDF_AT43_1   
35353261   Z3MKD_LWSK_EDDF_20170901130000  LWSK_EDDF_AT43  LWSK_EDDF_AT43_1   

              icao  dep_time  time_to_sector  entry          sector_hours  \
18        UKLVJEWH      1090              20   1110  UKLVJEWH_1800_1900_E   
43          UKBVSO      1090              53   1143    UKBVSO_1800_1900_E   
44          UKBVSO      1090              53   1143    UKBVSO_1900_2000_E   
67        UKBVTC12      1090              73   1163  UKBVTC12_1800_1900_E   
68        UKBVTC12      1090              73   1163  UKBVTC12_1900_2000_E   
...            ...       ...             ...    ...                   ...   
35353208   EDMMALB       785              97    882   EDMMALB_1500_1600_E   
35353227   EDGGKTG       785             104    889   EDGGKTG_1400_1500_E   
35353228   EDGGKTG       785             104    889   EDGGKTG_1500_1600_E   
35353260   EDGGPSA       785             112    897   EDGGPSA_1400_1500_E   
35353261   EDGGPSA       785             112    897   EDGGPSA_1500_1600_E   

         role  minute_start  minute_end  capacity  number   bool  overlap  \
18          E        1080.0      1139.0      39.0       2   True     60.0   
43          E        1080.0      1139.0      36.0       2   True     27.0   
44          E        1140.0      1199.0      36.0       2  False     33.0   
67          E        1080.0      1139.0      34.0       2   True      7.0   
68          E        1140.0      1199.0      34.0       2  False     53.0   
...       ...           ...         ...       ...     ...    ...      ...   
35353208    E         900.0       959.0      46.0       1  False     12.0   
35353227    E         840.0       899.0      41.0       1   True     41.0   
35353228    E         900.0       959.0      41.0       1  False     19.0   
35353260    E         840.0       899.0      38.0       1   True     33.0   
35353261    E         900.0       959.0      38.0       1  False     27.0   

              prob  weighted  
18        1.000000  0.500000  
43        0.450000  0.225000  
44        0.550000  0.275000  
67        0.116667  0.058333  
68        0.883333  0.441667  
...            ...       ...  
35353208  0.200000  0.200000  
35353227  0.683333  0.683333  
35353228  0.316667  0.316667  
35353260  0.550000  0.550000  
35353261  0.450000  0.450000  

[2882095 rows x 17 columns]

In [85]:
clean2 = clean.groupby('sector_hours')['weighted'].sum()

clean2 = pd.DataFrame({'sector_hours':clean2.index, 'Exp_freq':clean2.values})

clean2

sector_hours  Exp_freq
0      BIRDES_0000_0100_E  0.208333
1      BIRDES_0100_0200_E  9.009712
2      BIRDES_0200_0300_E  9.673326
3      BIRDES_0300_0400_E  1.572567
4      BIRDES_0400_0500_E  0.202083
...                   ...       ...
13336  UMMVS9_1900_2000_E  1.131944
13337  UMMVS9_2000_2100_E  0.538796
13338  UMMVS9_2100_2200_E  0.062885
13339  UMMVS9_2200_2300_E  0.497222
13340  UMMVS9_2300_2400_E  1.763889

[13341 rows x 2 columns]

In [86]:
sectorhours2 = sectorhours.merge(clean2, how='left', on='sector_hours')
sectorhours2

sector_hours      icao role  minute_start  minute_end  \
0       EGTTOCK_0000_0100_E   EGTTOCK    E             0          59   
1       EGTTOCK_0100_0200_E   EGTTOCK    E            60         119   
2       EGTTOCK_0200_0300_E   EGTTOCK    E           120         179   
3       EGTTOCK_0300_0400_E   EGTTOCK    E           180         239   
4       EGTTOCK_0400_0500_E   EGTTOCK    E           240         299   
...                     ...       ...  ...           ...         ...   
13485   LSAGL67_2010_2040_E   LSAGL67    E          1210        1239   
13486     LYBAL_1659_1800_E     LYBAL    E          1019        1079   
13487    LYBATN_1600_1700_E    LYBATN    E           960        1019   
13488    LYBAUN_1600_1700_E    LYBAUN    E           960        1019   
13489  LYBAUWES_1659_1800_E  LYBAUWES    E          1019        1079   

       capacity   Exp_freq  
0            49   4.403434  
1            49   2.524995  
2            49   1.470370  
3            49   1.245278  
4            49   2.629646  
...         ...        ...  
13485        23  12.153147  
13486        41  28.196892  
13487        43  27.533767  
13488        43  23.597230  
13489        54  44.990586  

[13490 rows x 7 columns]

In [88]:
sectorhours2['load_factor'] = sectorhours2['Exp_freq']/sectorhours2['capacity']
sectorhours2 = sectorhours2[['sector_hours', 'icao', 'minute_start', 'minute_end', 'Exp_freq', 'capacity', 'load_factor']]
sectorhours2

sector_hours      icao  minute_start  minute_end   Exp_freq  \
0       EGTTOCK_0000_0100_E   EGTTOCK             0          59   4.403434   
1       EGTTOCK_0100_0200_E   EGTTOCK            60         119   2.524995   
2       EGTTOCK_0200_0300_E   EGTTOCK           120         179   1.470370   
3       EGTTOCK_0300_0400_E   EGTTOCK           180         239   1.245278   
4       EGTTOCK_0400_0500_E   EGTTOCK           240         299   2.629646   
...                     ...       ...           ...         ...        ...   
13485   LSAGL67_2010_2040_E   LSAGL67          1210        1239  12.153147   
13486     LYBAL_1659_1800_E     LYBAL          1019        1079  28.196892   
13487    LYBATN_1600_1700_E    LYBATN           960        1019  27.533767   
13488    LYBAUN_1600_1700_E    LYBAUN           960        1019  23.597230   
13489  LYBAUWES_1659_1800_E  LYBAUWES          1019        1079  44.990586   

       capacity  load_factor  
0            49     0.089866  
1            49     0.051531  
2            49     0.030008  
3            49     0.025414  
4            49     0.053666  
...         ...          ...  
13485        23     0.528398  
13486        41     0.687729  
13487        43     0.640320  
13488        43     0.548773  
13489        54     0.833159  

[13490 rows x 7 columns]

In [90]:
sectorhours2.to_csv('C:\\Users\\utente\\Documents\\DSSC_IIyear\\Voli\\Dataframe\\shLF_exp.csv', index=None)

In [108]:
#sectorhours2[sectorhours2['Exp_freq']<1]
#clean[clean['sector_hours']=='EKDKY_0500_0520_E']

sector_hours     icao  minute_start  minute_end  Exp_freq  \
45     EISNLOW_0100_0200_E  EISNLOW            60         119  0.192550   
46     EISNLOW_0200_0300_E  EISNLOW           120         179  0.670163   
47     EISNLOW_0300_0400_E  EISNLOW           180         239  0.864583   
74       EKDKY_0440_0500_E    EKDKY           280         299  0.116667   
75       EKDKY_0500_0520_E    EKDKY           300         319  0.483333   
...                    ...      ...           ...         ...       ...   
13194   UMMVS9_0000_0100_E   UMMVS9             0          59  0.898148   
13214   UMMVS9_2000_2100_E   UMMVS9          1200        1259  0.538796   
13215   UMMVS9_2100_2200_E   UMMVS9          1260        1319  0.062885   
13216   UMMVS9_2200_2300_E   UMMVS9          1320        1379  0.497222   
13348  EHAACOD_0000_0100_E  EHAACOD             0          59  0.002083   

       capacity  load_factor  
45           20     0.009628  
46           20     0.033508  
47           20     0.043229  
74            4     0.029167  
75            4     0.120833  
...         ...          ...  
13194        46     0.019525  
13214        46     0.011713  
13215        46     0.001367  
13216        46     0.010809  
13348        44     0.000047  

[345 rows x 7 columns]

# 18    
    
    Con l'istante di partenza fissato, era necessario che:
    1) entry - start >= 0   ovvero:   start - entry <= 0     ovvero: start <= entry
    2) end - entry >= 0     ovvero:   end >= entry           ovvero: entry <= end
    
    Con la flessibilità di più/meno 30 minuti, basta che:
    1) (entry + 30) - start >= 0     ovvero: start - 30 <= entry     ovvero: start <= entry + 30
    2) end - (entry - 30) >= 0       ovvero: end + 30 > entry        ovvero: entry <= end + 30
    
    -----
    Esempio:

    A) settore-ora attivo dal min 60 al min 120
    B) settore-ora attivo dal min 0 al min 60
    C) settore-ora attivo dal min 100 al min 160

    Volo deve transitare per il settore al min 80 (fissato):
    Settore-ora A) OK
    Settore-ora B) NO
    Settore-ora C) NO
    
    Anticipando di 30 minuti (min 50):
    Settore-ora A) NO
    Settore-ora B) OK
    Settore-ora C) NO
    
    Posticipando di 30 minuti (min 110):
    Settore-ora A) OK
    Settore-ora B) NO
    Settore-ora C) OK